src/proof_production/UnsatCertificateNode.cpp

```diff
@@ -0,0 +1,498 @@
+/*********************                                                        */
+/*! \file UnsatCertificateNode.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Omri Isac, Guy Katz
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2022 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#include <Options.h>
+#include "UnsatCertificateNode.h"
+
+
+UnsatCertificateNode::UnsatCertificateNode( Vector<Vector<double>> *initialTableau, Vector<double> &groundUpperBounds, Vector<double> &groundLowerBounds )
+    : _parent( NULL )
+    , _contradiction( NULL )
+    , _hasSATSolution( false )
+    , _wasVisited( false )
+    , _delegationStatus( DelegationStatus::DONT_DELEGATE )
+    , _delegationNumber( 0 )
+    , _initialTableau( initialTableau )
+    , _groundUpperBounds( groundUpperBounds )
+    , _groundLowerBounds( groundLowerBounds )
+{
+}
+
+UnsatCertificateNode::UnsatCertificateNode( UnsatCertificateNode *parent, PiecewiseLinearCaseSplit split )
+    : _parent( parent )
+    , _contradiction( NULL )
+    , _headSplit( std::move( split ) )
+    , _hasSATSolution( false )
+    , _wasVisited( false )
+    , _delegationStatus( DelegationStatus::DONT_DELEGATE )
+    , _delegationNumber( 0 )
+    , _initialTableau( NULL )
+    , _groundUpperBounds( 0 )
+    , _groundLowerBounds( 0 )
+{
+    parent->_children.append( this );
+}
+
+UnsatCertificateNode::~UnsatCertificateNode()
+{
+    for ( auto *child : _children )
+    {
+        if ( child )
+        {
+            delete child;
+            child = NULL;
+        }
+    }
+
+    _children.clear();
+    deletePLCExplanations();
+
+    if ( _contradiction )
+    {
+        delete _contradiction;
+        _contradiction = NULL;
+    }
+
+    _parent = NULL;
+}
+
+void UnsatCertificateNode::setContradiction( Contradiction *contradiction )
+{
+    _contradiction = contradiction;
+}
+
+Contradiction *UnsatCertificateNode::getContradiction() const
```

Comment:
This is a bit tricky. The function is const, but it returns a pointer to an internal member of the class, which can then be changed from the outside. Better make the return value a "const Contraction *", so that that the caller can't change it.